sgkit/stats/grm.py

```diff
@@ -147,3 +149,655 @@ def genomic_relationship(
+    ...     relationship="stat_pedigree_relationship",
+    ... ).stat_inverse_relationship.values  # doctest: +NORMALIZE_WHITESPACE
+    array([[ 1.5,  0.5,  0. , -1. ,  0. ,  0. ],
+           [ 0.5,  2. ,  0.5, -1. , -1. ,  0. ],
+           [ 0. ,  0.5,  1.5,  0. , -1. ,  0. ],
+           [-1. , -1. ,  0. ,  2.5,  0.5, -1. ],
+           [ 0. , -1. , -1. ,  0.5,  2.5, -1. ],
+           [ 0. ,  0. ,  0. , -1. , -1. ,  2. ]])
+    >>> # inverse of a sub-matrix
+    >>> ds["subset_sample"] = "samples", [False, False, False, True, True, True]
+    >>> sg.invert_relationship_matrix(
+    ...     ds,
+    ...     relationship="stat_pedigree_relationship",
+    ...     subset_sample="subset_sample",
+    ... ).stat_inverse_relationship.values.round(3)  # doctest: +NORMALIZE_WHITESPACE
+    array([[   nan,    nan,    nan,    nan,    nan,    nan],
+           [   nan,    nan,    nan,    nan,    nan,    nan],
+           [   nan,    nan,    nan,    nan,    nan,    nan],
+           [   nan,    nan,    nan,  1.567,  0.233, -1.   ],
+           [   nan,    nan,    nan,  0.233,  1.567, -1.   ],
+           [   nan,    nan,    nan, -1.   , -1.   ,  2.   ]])
+    """
+    A = da.array(ds[relationship].data)
+    if subset_sample is None:
+        Ainv = da.linalg.inv(A)
+    else:
+        idx = ds[subset_sample].values
+        Ainv = _sub_matrix_inv(A, idx, chunks=subset_rechunk)
+    new_ds = create_dataset(
+        {
+            variables.stat_inverse_relationship: (
+                ("samples_0", "samples_1"),
+                Ainv,
+            )
+        }
+    )
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+def _hybrid_full_matrix_Martini(
+    Ainv: ArrayLike, G: ArrayLike, tau: float = 1, omega: float = 1
+):
+    # calculate sub-matrices of H
+    if (tau == 1) and (omega == 1):
+        H = G
+    elif omega == 1:
+        H = 1 / tau * G
+        # H = da.linalg.inv(tau * da.linalg.inv(G))
+    else:
+        H = da.linalg.inv(tau * da.linalg.inv(G) + (1 - omega) * Ainv)
+    return H
+
+
+def _hybrid_sub_matrix_Martini(
+    A: ArrayLike,
+    G: ArrayLike,
+    is_genomic: ArrayLike,
+    tau: float = 1,
+    omega: float = 1,
+    A22inv: Optional[ArrayLike] = None,
+):
+    # Note: `is_genomic` must be a numpy array to determine matrix sizes
+    not_genomic = ~is_genomic  # indices of non G-matrix
+    # get sub-matrices
+    G22 = G[is_genomic, :][:, is_genomic]
+    A11 = A[not_genomic, :][:, not_genomic]
+    A12 = A[not_genomic, :][:, is_genomic]
+    A21 = A[is_genomic, :][:, not_genomic]
+    A22 = A[is_genomic, :][:, is_genomic]
+    # optional inversions required for consistent API
+    if A22inv is None:
+        A22inv = da.linalg.inv(A22)
+    else:
+        A22inv = A22inv[is_genomic, :][:, is_genomic]
+    if (tau == 1) and (omega == 1):
+        H22 = G22
+    elif omega == 1:
+        H22 = 1 / tau * G22
+    else:
+        H22 = da.linalg.inv(tau * da.linalg.inv(G22) + (1 - omega) * A22inv)
```

Comment:
This line looks like it could involve a lot of computation (an inverse of another inverse). Worth splitting across two lines, just for the sake of providing accurate feedback on line-numbers when doing profiling?